core/modules/replica/JobController.cc

```diff
@@ -0,0 +1,503 @@
+    //   calling the thread-safe methods. A problem is that until they finish
+    //   in a clean way the thread will never finish, and the application will
+    //   hang on _thread->join().
+
+    // util::Lock lock(_mtx)  (disabled)
+
+    if (not isRunning()) return;
+
+    // Tell the thread to finish
+    _stop = true;
+
+    // Join with the thread before clearning up the pointer
+    _thread->join();
+    _thread.reset(nullptr);
+
+    _stop = false;
+}
+
+void JobController::join() {
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::join");
+    if (_thread) _thread->join();
+}
+
+FindAllJob::Ptr JobController::findAll(std::string const& databaseFamily,
+                                       bool saveReplicaInfo,
+                                       FindAllJob::CallbackType onFinish,
+                                       Job::Options const& options) {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::findAll");
+
+    util::Lock lock(_mtx, "JobController::findAll");
+
+    auto const self = shared_from_this();
+    auto const job =
+        FindAllJob::create(
+            databaseFamily,
+            saveReplicaInfo,
+            _controller,
+            std::string(),
+            [self] (FindAllJob::Ptr job) {
+                self->onFinish (job);
+            },
+            options
+        );
+
+    // Register the job (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+
+    _registry[job->id()] =
+        std::make_shared<JobWrapperImpl<FindAllJob>>(job, onFinish);
+
+    // Initiate the job
+    //
+    // FIXME: don't start the job right away. Put the request into the priority queue
+    // and call the scheduler's method to evaluate jobs in the queue to
+    // to see which should be started next (if any).
+
+    job->start();
+
+    return job;
+}
+
+FixUpJob::Ptr JobController::fixUp(std::string const& databaseFamily,
+                                   FixUpJob::CallbackType onFinish,
+                                   Job::Options const& options) {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::fixUp");
+
+    util::Lock lock(_mtx, "JobController::fixUp");
+
+    auto const self = shared_from_this();
+    auto const job =
+        FixUpJob::create(
+            databaseFamily,
+            _controller,
+            std::string(),
+            [self] (FixUpJob::Ptr job) {
+                self->onFinish (job);
+            },
+            options
+        );
+
+    // Register the job (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+
+    _registry[job->id()] =
+        std::make_shared<JobWrapperImpl<FixUpJob>>(job, onFinish);
+
+    // Initiate the job
+    //
+    // FIXME: don't start the job right away. Put the request into the priority queue
```

Comment:
ticket number?

Reply:
The JobController has been redesigned. It's much simpler now. That FIXME is gone.